framework/src/loops/MaxQpsThread.C

```diff
@@ -47,47 +47,21 @@ MaxQpsThread::operator()(const ConstElemRange & range)
     // This ensures we can access the correct qrules if any block-specific
     // qrules have been created.
     assem.setCurrentSubdomainID(elem->subdomain_id());
+    assem.setVolumeQRule(elem);
 
-    FEType fe_type(FIRST, LAGRANGE);
-    unsigned int dim = elem->dim();
-    unsigned int side = 0; // we assume that any element will have at least one side ;)
-
-    // We cannot mess with the FE objects in Assembly, because we might need to request second
-    // derivatives
-    // later on. If we used them, we'd call reinit on them, thus making the call to request second
-    // derivatives harmful (i.e. leading to segfaults/asserts). Thus, we have to use a locally
-    // allocated object here.
-    //
-    // We'll use one for element interiors, which calculates nothing
-    std::unique_ptr<FEBase> fe(FEBase::build(dim, fe_type));
-    fe->get_nothing();
-    // Alex: This seems wrong, e.g. like we'd want the quadrature order to be higher if the
-    // polynomial basis is higher, but our fe_type up above is first Lagrange, so it seems like the
-    // person who wrote this code didn't want to consider higher order bases even if if the user
-    // asked for higher order bases
-    fe->add_p_level_in_reinit(false);
-
-    // And another for element sides, which calculates the minimum
-    // libMesh currently allows for that
-    std::unique_ptr<FEBase> side_fe(FEBase::build(dim, fe_type));
-    side_fe->get_xyz();
-    side_fe->add_p_level_in_reinit(false);
-
-    // figure out the number of qps for volume
-    auto qrule = assem.attachQRuleElem(dim, *fe);
-    fe->reinit(elem);
+    auto & qrule = assem.writeableQRule();
+    qrule->init(elem->type(), elem->p_level());
     if (qrule->n_points() > _max)
       _max = qrule->n_points();
 
-    // figure out the number of qps for the face
-    // NOTE: user might specify higher order rule for faces, thus possibly ending up with more qps
-    // than in the volume
-    auto qrule_face = assem.attachQRuleFace(dim, *side_fe);
-    if (dim > 0) // side reinit in 0D makes no sense, but we may have NodeElems
+    if (elem->dim())
     {
-      side_fe->reinit(elem, side);
+      assem.setFaceQRule(elem, /*side=*/0);
+      const std::unique_ptr<const Elem> side(elem->build_side_ptr(0));
```

Comment:
I don' think we need `side` here - `side->p_level()` is just going to be `elem->p_level()`, and `side->type()` is just going to be `elem->side_type(0)`.

But this code is correct and I wouldn't stress about a little extra heap manipulation here.

Reply:
Perhaps we should add a comment here, about assuming that side 0 always ends up with at least as many quadrature points as any other side?  That's *true*, since we number the quad side of a pyramid before the triangle sides and even the worst case triangle quadrature rules are no more expensive than a quad rule of the same order, but it's a cute benefit of libMesh numbering convention, not a priori true.

Reply:
> elem->side_type(0)

Did not know about this API!